packages/shared/src/components/modals/post/CreateSharedPostModal.tsx

```diff
@@ -0,0 +1,125 @@
+import React, { FormEventHandler, ReactElement, useRef, useState } from 'react';
+import { Modal, ModalProps } from '../common/Modal';
+import { ExternalLinkPreview } from '../../../graphql/posts';
+import MarkdownInput, { MarkdownRef } from '../../fields/MarkdownInput';
+import { WriteLinkPreview, WritePreviewSkeleton } from '../../post/write';
+import useDebounce from '../../../hooks/useDebounce';
+import { isValidHttpUrl } from '../../../lib/links';
+import { usePostToSquad } from '../../../hooks';
+import { Button, ButtonSize } from '../../buttons/Button';
+import AtIcon from '../../icons/At';
+import { Divider } from '../../utilities';
+import SourceButton from '../../cards/SourceButton';
+import { Squad } from '../../../graphql/sources';
+import { formToJson } from '../../../lib/form';
+
+interface CreateSharedPostModalProps extends ModalProps {
+  preview: ExternalLinkPreview;
+  onSharedSuccessfully?: () => void;
+  squad: Squad;
+}
+
+export function CreateSharedPostModal({
+  preview,
+  squad,
+  onSharedSuccessfully,
+  ...props
+}: CreateSharedPostModalProps): ReactElement {
+  const markdownRef = useRef<MarkdownRef>();
+  const [link, setLink] = useState(preview?.permalink ?? preview?.url ?? '');
+  const {
+    getLinkPreview,
+    isLoadingPreview,
+    preview: updatedPreview,
+    isPosting,
+    onSubmitPost,
+  } = usePostToSquad({
+    initialPreview: preview,
+    onPostSuccess: () => {
+      if (onSharedSuccessfully) onSharedSuccessfully();
+      props.onRequestClose(null);
+    },
+  });
+  const onFormSubmit: FormEventHandler<HTMLFormElement> = (e) => {
+    e.preventDefault();
+
+    const { commentary } = formToJson(e.currentTarget);
+
+    return onSubmitPost(e, squad.id, commentary);
+  };
+
+  const [checkUrl] = useDebounce((value: string) => {
+    const links = [link, updatedPreview?.url, updatedPreview?.permalink];
+    if (!isValidHttpUrl(value) || links.some((url) => url === value))
+      return null;
+
+    return getLinkPreview(value);
+  }, 1000);
```

Comment:
We are using this 3 times now I see, perhaps we can simply it in a custom hook to keep it aligned if anything ever changes

Reply:
Good point. I have created a hook for this now.